File: Editorial/2_Add_Two_Numbers/CSharp/2_Recursive_Addition.cs
public class Solution {
    public ListNode AddTwoNumbers(ListNode l1, ListNode l2) {
        return Add(l1, l2, 0);
    }

    private ListNode Add(ListNode l1, ListNode l2, int carry) {
        if (l1 == null && l2 == null && carry == 0) {
            return null;
        }

        int total = carry;
        if (l1 != null) {
            total += l1.val;
            l1 = l1.next;
        }

        if (l2 != null) {
            total += l2.val;
            l2 = l2.next;
        }

        ListNode node = new ListNode(total % 10);
        node.next = Add(l1, l2, total / 10);
        return node;
    }
}

File: Editorial/2_Add_Two_Numbers/Go/2_Recursive_Addition.go
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    return add(l1, l2, 0)
}

func add(l1 *ListNode, l2 *ListNode, carry int) *ListNode {
    if l1 == nil && l2 == nil && carry == 0 {
        return nil
    }

    total := carry
    if l1 != nil {
        total += l1.Val
        l1 = l1.Next
    }

    if l2 != nil {
        total += l2.Val
        l2 = l2.Next
    }

    return &ListNode{
        Val:  total % 10,
        Next: add(l1, l2, total/10),
    }
}
